index.html

```diff
@@ -346,8 +352,9 @@ <h2>Suite Definition</h2>
       
       <p>
         <code>TypedData</code> MUST be a JSON object according to the EIP712 specification and
-        contains properties <code>types</code>, <code>domain</code>, <code>primaryType</code> and
-        <code>message</code>. 
+      contains properties <code>types</code>, <code>domain</code>, <code>primaryType</code> and
+      <code>message</code>. <code>TypedData</code> can be generated by the <a href="#dfn-schema-generation-algorithm">schema generation algorithm</a> 
+      if not provided as input.
```

Comment:
Agree with the `primaryType` nuances. 
Don't fully understand how we can generate a value for `domain` as that is implementation-specific on what the developer wants to add. AFAIK, each individual property in `EIP712Domain` is optional, but *something* is required. I believe this should be passed down by the developer.

Reply:
The changes for `primaryType` look good to me.

Allowing the user/developer to set the properties of `domain` (the instance of EIP712Domain) is reasonable.

The changes here are still a little inconsistent with the rest of the document:
> TypedData can be generated by the schema generation algorithm if not provided as input.

TypedData is never provided as input; it is generated internally as part of this signature suite, with the following properties:
- `message` is derived from the linked data document and proof object to be signed.
- `primaryType`, `domain`, and `types` are taken from the `eip712Domain` property of the proof object (which is suggested to be renamed to `eip712` in #5)

The types generation algorithm added in this PR says it generates `TypedData`, but it actually generates a `types` map, that is one property of `TypedData`. I've suggested changes to fix this:
- https://github.com/w3c-ccg/ethereum-eip712-signature-2021-spec/pull/25/files#r699407907
- https://github.com/w3c-ccg/ethereum-eip712-signature-2021-spec/pull/25/files#r699408191
- https://github.com/w3c-ccg/ethereum-eip712-signature-2021-spec/pull/25/files#r699408365

Reply:
I've addressed the changes you suggested, thanks for all of those. 

I'm confused about "`primaryType`, `domain`, and `types` are taken from the `eip712Domain` property of the proof object"

From my understanding, at least according to the current test vectors, the input document's `proof` object is bare minimum specifying only a few fields, none of which have to do with `eip712Domain`. Are you suggesting that if we leave it up to the developer to specify these fields, they should put them in the `proof` document before creating a proof through the suite? 

My thought was to have them be optional properties in the signature suite constructor and leave the input document as-is. I would prefer having them in the `proof` object prior to signing over the document to keep the constructor cleaner, but want to ensure everyone is on the same page.

Reply:
LD Proofs specifies in [Proof Algorithm](https://w3c-ccg.github.io/ld-proofs/#proof-algorithm) that when creating a proof object, *proof options* are taken as input and passed into the proof object. I would expect the input document for signing to not have a proof object, or if it does then it is removed first, as in [Proof Verification Algorithm](https://w3c-ccg.github.io/ld-proofs/#proof-verification-algorithm).

What we current specify as the proof's `eip712Domain` property (containing properties `primaryType`, `domain`, and `types` - possibly omitting `types` and `primaryType` if using the proposed types generation algorithm) would be passed in the *proof options* for signing. The signature suite constructs a proof object given the *proof options*, picking defaults for required properties as needed. But the `eip712Domain` property passed from *proof options* would not be put into the proof object at this point. Instead, its properties `primaryType`, `domain`, and `types` (generating `types` and `primaryType` using the types generation algorithm if needed), would be put in the `TypedData` object. The remaining property of the `TypedData` object, `message`, would have as its value the input document with single proof object inserted. After signing, the signature (`proofValue`) and `eip712Domain` object are then inserted into the proof object. This should be what the current test vectors should demonstrate. However the test vectors currently do not show the input proof options and document that would result in the TypedData object in Example 5.

So the proof object appears without the `eip712Domain` property in the TypedData object in Examples 5 and 6 because the `eip712Domain` property values are instead in the other `TypedData` properties outside `message`. I don't think the `types` property can be in the `message` anyway, because then there would need to be additional `types` in `TypedData` to cover it, and I don't know if that would be possible to do consistently. I don't think it creates a signature malleability problem to have these properties placed in the proof object after signing, since the types and domain are included in the EIP-712 encoding and signing process.